internal/charmstore/resources.go

```diff
@@ -0,0 +1,244 @@
+		query = append(query, bson.DocElem{"revision", revision})
+	}
+	return query
+}
+
+// sortResources sorts the provided resource docs, The resources are
+// sorted first by URL then by name and finally by revision.
+func sortResources(resources []*mongodoc.Resource) {
+	sort.Sort(resourcesByName(resources))
+}
+
+type resourcesByName []*mongodoc.Resource
+
+func (sorted resourcesByName) Len() int      { return len(sorted) }
+func (sorted resourcesByName) Swap(i, j int) { sorted[i], sorted[j] = sorted[j], sorted[i] }
+func (sorted resourcesByName) Less(i, j int) bool {
+	r0, r1 := sorted[i], sorted[j]
+	if *r0.BaseURL != *r1.BaseURL {
+		return r0.BaseURL.String() < r1.BaseURL.String()
+	}
+	if r0.Name != r1.Name {
+		return r0.Name < r1.Name
+	}
+	return r0.Revision < r1.Revision
+}
+
+var resourceNotFound = errgo.Newf("resource not found")
+
+// ListResources returns the set of resources for the charm. If the
+// unpublished channel is specified then set is composed of the latest
+// revision for each resource. Otherwise it holds the revisions declared
+// when the charm/channel pair was published.
+func (s *Store) ListResources(entity *mongodoc.Entity, channel params.Channel) ([]*mongodoc.Resource, error) {
+	if entity.URL.Series == "bundle" {
+		return nil, nil
+	}
+	if entity.CharmMeta == nil {
+		return nil, errgo.Newf("entity missing charm metadata")
+	}
+	baseEntity, err := s.FindBaseEntity(entity.URL, FieldSelector("channelresources"))
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	// get all of the resources associated with the charm first.
+	resources, revisions, err := s.charmResources(entity.BaseURL)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	if channel != params.UnpublishedChannel {
+		revisions = mapRevisions(baseEntity.ChannelResources[channel])
+	}
+	var docs []*mongodoc.Resource
+	for name := range entity.CharmMeta.Resources {
+		revision, ok := revisions[name]
+		if !ok {
+			// There is no matching resource available, perhaps this is an error?
+			continue
+		}
+		doc := resources[name][revision]
+		if doc == nil {
+			return nil, errgo.Newf("published resource %q, revision %d not found", name, revision)
+		}
+		docs = append(docs, doc)
+	}
+	sortResources(docs)
+	return docs, nil
+}
+
+// charmResources returns all of the currently stored resources for a charm.
+func (s *Store) charmResources(baseURL *charm.URL) (map[string]map[int]*mongodoc.Resource, map[string]int, error) {
+	resources := make(map[string]map[int]*mongodoc.Resource)
+	latest := make(map[string]int)
+	iter := s.DB.Resources().Find(bson.D{{"baseurl", baseURL}}).Iter()
+	var r mongodoc.Resource
+	for iter.Next(&r) {
+		resource := new(mongodoc.Resource)
+		*resource = r
+		if _, ok := resources[r.Name]; !ok {
+			resources[r.Name] = make(map[int]*mongodoc.Resource)
+		}
+		resources[r.Name][r.Revision] = resource
+		if r.Revision >= latest[r.Name] {
+			latest[r.Name] = r.Revision
+		}
+	}
+	return resources, latest, iter.Close()
```

Comment:
I think it would be better to return zero values if the iterator fails.

    if err := iter.Close(); err != nil {
        return errgo.Mask(err)
    }